File: dev-c++/0008-string-to-integer-atoi.cpp
#include <numeric>


class Solution {
public:
    int myAtoi(string s) {
        bool leading = true;
        bool negative = false;
        long result = 0;

        for(int i = 0; i < s.size(); ++ i) {
            if(leading && s[i] == ' ') {
                continue;
            }

            if(leading && s[i] == '+') {
                leading = false;
                continue;
            }

            if(leading && s[i] == '-') {
                leading = false;
                negative = true;
                continue;
            }

            if(s[i] >= '0' && s[i] <= '9') {
                leading = false;
                result *= 10;

                if (negative) { result -= (s[i] - '0'); }
                else          { result += (s[i] - '0'); }   

                if(result > INT_MAX) { return INT_MAX; }
                if(result < INT_MIN) { return INT_MIN; }

                continue;                
            }
            
            break;
        }

        return (int) result;
    }
};

File: dev-rust/0008-string-to-integer-atoi.rs
impl Solution {
    pub fn my_atoi(s: String) -> i32 {
        let mut leading  = true;
        let mut negative = false;
        let mut result: i32 = 0;

        for ch in s.chars() { 
            if leading && ch == ' ' { continue; }

            if leading && (ch == '+' || ch == '-') {
                if ch == '-' { negative = true; }

                leading = false;
                continue;
            }

            if ch >= '0' && ch <= '9' {
                let digit: i32 = (ch as i32) - 0x30;
                
                leading = false;
                result  = result.saturating_mul(10);

                if negative { result = result.saturating_sub(digit); }
                else        { result = result.saturating_add(digit); }     

                continue;
            }

            break;
        }

        return result ;
    }
}
